cocotb/scheduler.py

```diff
@@ -876,31 +876,25 @@ def _schedule(self, coroutine, trigger=None):
             if _debug:
                 self.log.debug(f"Scheduling with {send_outcome}")
 
-            coro_completed = False
-            try:
-                coroutine._trigger = None
-                result = coroutine._advance(send_outcome)
-                if _debug:
-                    self.log.debug("Coroutine %s yielded %s (mode %d)" %
-                                   (coroutine._coro.__qualname__, str(result), self._mode))
+            coroutine._trigger = None
+            result = coroutine._advance(send_outcome)
 
-            except cocotb.decorators.CoroutineComplete:
+            if coroutine._finished:
                 if _debug:
                     self.log.debug("Coroutine {} completed with {}".format(
                         coroutine, coroutine._outcome
                     ))
-                coro_completed = True
-
-            # this can't go in the else above, as that causes unwanted exception
-            # chaining
-            if coro_completed:
+                assert result is None
                 self._unschedule(coroutine)
 
             # Don't handle the result if we're shutting down
             if self._terminate:
                 return
 
-            if not coro_completed:
+            if not coroutine._finished:
+                if _debug:
+                    self.log.debug("Coroutine %s yielded %s (mode %d)" %
+                                   (coroutine._coro.__qualname__, str(result), self._mode))
```

Comment:
Are we sure that `str(result)` is always going to result in something reasonable?

Reply:
Doubtful. The debug logging needs some improvements.

Reply:
We can leave it for another PR?

Reply:
We can. I was also thinking it could be moved to after the `_trigger_from_any` call, so you know what will be printed will be a trigger. Either way works for me, you have the checkmark.